### DISCUSSION TITLE
Better hit detection with symbol layer icon

### DISCUSSION THREAD

**Discussion by qpincon** [maplibre/maplibre-gl-js#3978] [Q&A]
Hello,

I have a symbol layer with a custom SDF icon in it. I am listening for mouse events on this layer to detect hover, to change the halo color and width.

The thing is, hit testing happens using the bounding box of the displayed image. This is a bit problematic when the symbol is not rectangular, or if it is rotated.

It's possible to add a function that will improve hit detection to accommodate the rotation, but I was wondering if there are some existing method / doc to do this kind of things. Also, I guess with SDF it could be possible to enable precise hit testing depending on the shape directly.

Thank you for you help, awesome library!

**Comment by HarelM** [comment]
@kubapelc is working on improving hit detection due to the globe work he is doing.
But it's not by improving the bounding box to be the symbol itself.
I think moving from a rectangle (let's ignore pitch for a minute here) to a shape can cause performance degradation when it comes to collision detection of labels and symbols, so going in that direction might not be a good approach.
Having said that, if you know how to improve that, and you can verify that collision detection is still happening very fast we can consider having a parameter to the hit detection to use precise shape or something, IDK.

If you want to pursue this I would recommend creating a short design document to align everyone on what you plan to achieve and how.

**Reply by qpincon** [reply]
Thank you for your answer.
I am not familiar at all with the codebase, but it's a possibility that I could contribute to the library in the future. Do you have resources on how to get started and could you point me where the hit detection code resides please?

On your remark: I guess precise hit detection could only be triggered on the matches of the current bounding box method. This was, the "coarse" filtering would still be fast, and the precise algorithm would filter out even more. But as I said I don't know the code at all so I don't know if it would be that "easy".

**Reply by HarelM** [reply]
Checkout:
https://github.com/maplibre/maplibre-gl-js/blob/main/CONTRIBUTING.md
There are a lot of readme file scattered throughout the repo for specific stuff, look for those when you try and run tests, compile etc.
Regarding collision detection,
Currently there's an open PR that addresses some of the issue, You can take a look at it and see if you understand some of the stuff it does.
 - #4071

There's the "Collision Index" class that is responsible for some of the collision logic, and other classes as well including symbol_draw etc.
